store/sqlstore/post_store.go

```diff
@@ -1312,42 +1312,73 @@ func (s *SqlPostStore) GetMaxPostSize() int {
 }
 
 func (s *SqlPostStore) GetParentsForExportAfter(limit int, afterId string) ([]*model.PostForExport, *model.AppError) {
-	var posts []*model.PostForExport
-	_, err := s.GetSearchReplica().Select(&posts, `
-                SELECT
-                    p1.*,
-                    Users.Username as Username,
-                    Teams.Name as TeamName,
-                    Channels.Name as ChannelName
-                FROM
-                    Posts p1
-                INNER JOIN
-                    Channels ON p1.ChannelId = Channels.Id
-                INNER JOIN
-                    Teams ON Channels.TeamId = Teams.Id
-                INNER JOIN
-                    Users ON p1.UserId = Users.Id
-                WHERE
-                    p1.Id > :AfterId
-                    AND p1.ParentId = ''
-                    AND p1.DeleteAt = 0
-					AND Channels.DeleteAt = 0
-					AND Teams.DeleteAt = 0
-                ORDER BY
-                    p1.Id
-                LIMIT
-                    :Limit`,
-		map[string]interface{}{"Limit": limit, "AfterId": afterId})
+	for {
+		var parentPosts []*model.Post
+		_, err := s.GetSearchReplica().Select(&parentPosts,
+			`SELECT
+				*
+			FROM
+				Posts
+			WHERE
+				Id > :AfterId
+				AND RootId = ''
+				AND DeleteAt = 0
+			ORDER BY Id
+			LIMIT :Limit`,
+			map[string]interface{}{"Limit": limit, "AfterId": afterId})
+		if err != nil {
+			return nil, model.NewAppError("SqlPostStore.GetAllAfterForExport", "store.sql_post.get_posts.app_error",
+				nil, err.Error(), http.StatusInternalServerError)
+		}
 
-	if err != nil {
-		return nil, model.NewAppError("SqlPostStore.GetAllAfterForExport", "store.sql_post.get_posts.app_error",
-			nil, err.Error(), http.StatusInternalServerError)
+		var postsForExport []*model.PostForExport
+		if len(parentPosts) == 0 {
+			return postsForExport, nil
+		}
+
+		rootIds := []string{}
+		for _, post := range parentPosts {
+			rootIds = append(rootIds, post.Id)
+		}
+		keys, params := MapStringsToQueryParams(rootIds, "PostId")
+
+		_, err = s.GetSearchReplica().Select(&postsForExport, `
+		SELECT
+			p1.*,
+			Users.Username as Username,
+			Teams.Name as TeamName,
+			Channels.Name as ChannelName
+		FROM
+			(Select * FROM Posts WHERE Id IN `+keys+`) p1
+		INNER JOIN
+			Channels ON p1.ChannelId = Channels.Id
+		INNER JOIN
+			Teams ON Channels.TeamId = Teams.Id
+		INNER JOIN
+			Users ON p1.UserId = Users.Id
+		WHERE
+			Channels.DeleteAt = 0
```

Comment:
Slightly off topic, but it's unfortunate our export strategy looks at posts outside of the channels or team context. Few of our indexes are written to accommodate team-free/channel-free queries like this, which might explain a bit of the pain.

I suspect this query could be simplified (likely not requiring a `for` loop) if we filtered to a team and channel and iterated through the teams and channels in the export code instead.

Reply:
I agree with the logic here, and did spend some time on that before deciding to just push the improvement I had and then making further improvements in a follow-up PR.

How does it sound if I capture this idea, along with all the other notes I have, when I create a backlog ticket for a more thorough export overhaul?

Reply:
👍